### PR TITLE
Core Logic DRY Refactoring

### PR DESCRIPTION
**Pull request by kodinkat**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208361834831595

### CODE REVIEW

File: src/Controllers/Admin/AppSettingsController.php [hunk 1]
```diff
@@ -323,11 +239,25 @@ public function up( Request $request, $params ) {
...
+    /**
+     * Handle directional movement of apps within
+     * admin list.
+     *
+     * @param string $direction The direction (up/down).
+     * @param array $params The route parameters.
+     *
+     * @return ResponseInterface The RedirectResponse instance.
+     */
+    private function handle_up_down_direction_request( $direction, $params ) {
```
Comment (incraigulous) [inline src/Controllers/Admin/AppSettingsController.php:254]: Small change: let's rename this to "move"

**Review by incraigulous — CHANGES_REQUESTED** [review]

**Review by incraigulous — COMMENTED** [review]
This is a great step! The complexity will come from having three sources for apps, coded, custom (admin), custom (user), and then having one service to manage them all. We discussed this in the meeting after you logged off. 

I think we are going to need to have three different services to manage them all, when we need to work with them individually, and then one service when we need to work with them all together. They can all share a common base or interface to keep things sane. Let me see if I can get us pointed in a good direction.

**Comment by incraigulous** [comment]
@kodinkat I've pushed my app sources refactor. Here's a walkthrough: 

https://www.loom.com/share/a74661be797b4bf7bb1df2aab8813ee7

**Review by incraigulous — APPROVED** [review]

**Comment by incraigulous** [comment]
@kodinkat , great work! Your work 

 - I LOVED your idea of having `Trainings` using the Source API, but I also noticed that we had some issues around training not having slugs, so some of the existing `AppSources` methods didn't work.  To accommodate your idea, I broke a more generic [Source](https://github.com/thecodezone/dt-home/pull/196/commits/3a570a9152bd391f8ad56ffd6b73aa36152f0a23#diff-95ec3c4d24d4137cd1eb4d6e0f7d864a7814de0aec72e66aa787510cb390cbd0) abstract that is not some of the things we hard-coded for apps, like slugs and sort keys. AppSource is not extending Source to provide app-specific functionality. You can now use the `Trainings` source to find, update, create, etc, without issue.  
 - Again, I LOVED your `uber_sort` function which is a good improvement over our sort function, so I promoted it to replace our existing sort function. I [modified](https://github.com/thecodezone/dt-home/pull/196/commits/3a570a9152bd391f8ad56ffd6b73aa36152f0a23#diff-b84c9c1287c73cbced2ea3fd7995acce18e3567841e21c1ee94e741f47588c46R146) your uber sort calls to use the standard sort function, which now uses your code.
 - I ran out of time before I could fully test my work! Would you see look around and see if I seem to have broken anything? If not, I'll merge this PR!

**Comment by kodinkat** [comment]
> @kodinkat , great work!
> 
> * I LOVED your idea of having `Trainings` using the Source API, but I also noticed that we had some issues around the trainings objects not having slugs, so some of the existing `AppSources` methods didn't work.  To accommodate your idea, I broke a more generic [Source](https://github.com/thecodezone/dt-home/pull/196/commits/3a570a9152bd391f8ad56ffd6b73aa36152f0a23#diff-95ec3c4d24d4137cd1eb4d6e0f7d864a7814de0aec72e66aa787510cb390cbd0) abstract that generalizes some of the things we hard-coded for apps, like slugs and sort keys. AppSource is now extending Source to provide app-specific functionality. You can now use the `Trainings` source to find, update, create, etc, without issue.
> * Again, I LOVED your `uber_sort` function which is a good improvement over our sort function, so I promoted it to replace our existing sort function. I [modified](https://github.com/thecodezone/dt-home/pull/196/commits/3a570a9152bd391f8ad56ffd6b73aa36152f0a23#diff-b84c9c1287c73cbced2ea3fd7995acce18e3567841e21c1ee94e741f47588c46R146) your uber sort calls to use the standard sort function, which now uses your code.
> * I ran out of time before I could fully test my work! Would you see look around and see if I seem to have broken anything? If not, I'll merge this PR!

@incraigulous Good stuff! For sure, I'll have another nose around, to see if I can smoke out any more hidden bugs.. :)

**Pull request merged**